pkg/operator/k8sutil/events.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package k8sutil for Kubernetes helpers.
+package k8sutil
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+type eventObject struct {
+	eventsReportedAt time.Time
+	eventCount       int
+}
+
+// EventReporter is custom events reporter type which allows user to limit the events
+type EventReporter struct {
+	recorder       record.EventRecorder
+	reportedEvents map[string]*eventObject
+
+	// report events x times where x is count
+	count int
+
+	// report events after x minutes
+	eventReportAfterMinutes int
+}
+
+// NewEventReporter returns EventReporter object
+func NewEventReporter(recorder record.EventRecorder, maxCountInGivenTime, reportAfter int) *EventReporter {
+	er := &EventReporter{
+		recorder:                recorder,
+		count:                   maxCountInGivenTime,
+		eventReportAfterMinutes: reportAfter,
+	}
+
+	er.reportedEvents = map[string]*eventObject{}
+
+	return er
+}
+
+// Report records a events if eventReportAfterMinutes has passed or events occurred less than count
+func (rep *EventReporter) Report(instance runtime.Object, eventType, eventReason, msg string) {
```

Comment:
> It will be a little difficult to test this out without a k8s setup as we do need a `manager` for this and for getting a `manager` we do need a `kubeconfig` file. Pls, let me know any solution. Any pointer would be helpful.

K8s is mocked in all the rook unit tests, so we should be able to mock this as well. See existing tests such as [TestUpdateStatus](https://github.com/rook/rook/blob/22721f1df8cbd29dd7bb88cf1dbb26b18461c075/pkg/operator/ceph/controller/status_test.go#L30) that use a fake runtime.

Reply:
What about using https://pkg.go.dev/k8s.io/client-go/tools/record#FakeRecorder?

Reply:
With this fake recorder, we can not make sure that we have actually generated an event on the cluster. Because there is no way to get recorded events via a fake recorder.

As we did not found a better way to do that I am using this one only and making sure the events counts etc via variables in the objects I have created. If we do found a better way future we can add that also.

Reply:
@iamniting , you can get recorded events out of the fake recorder, here is one example of event tests: https://github.com/rohantmp/local-storage-operator/blob/eventReporter/pkg/common/events/event_reporter_test.go

However, it records events of strings, but you can check if a particular event has occured by converting it to the equivalent key `key := fmt.Sprintf("%s %s %s", ev.GetType(), ev.GetReason(), ev.GetMessage())`